### PR TITLE
Search in both filename and contents

### PR DESCRIPTION
**Pull request by harshad1**
In this PR I have added functionality for searching both contents _and_ title if so desired.

Also included are a few tweaks which should improve search performance.

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java [hunk 1]
```diff
@@ -129,13 +136,13 @@ public QueueSearchFilesTask(final SearchOptions config, final GsCallback.a1<List
                     }
                 }
             }
-            _regex = pattern;
+            _matcher = pattern != null ? pattern.matcher("") : null;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java:139]: resetting a matcher is slightly faster than creating one

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java [hunk 2]
```diff
@@ -198,49 +205,37 @@ protected List<FitFile> doInBackground(Void... voidp) {
...
+            final Queue<File> subQueue = new LinkedList<>();
 
             try {
                 if (!currentDir.canRead() || currentDir.isFile()) {
                     return subQueue;
                 }
 
-                File[] subDirsOrFiles = currentDir.listFiles();
+                final File[] subDirsOrFiles = currentDir.listFiles();
+                final int trimSize = _config.rootSearchDir.getCanonicalPath().length() + 1;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java:216]: Just trim the first n characters rather than find replace

**Comment by harshad1** [comment]
/home/harshad/Downloads/Screenshot_20230311-222704.png

Search options

**Comment by harshad1** [comment]
Results
image

**Comment by harshad1** [comment]
@adelobosko Any feedback you have here would be nice

**Comment by adelobosko** [comment]
> @adelobosko Any feedback you have here would be nice

Seems all ok. I like this PR.  It is much better.

**Comment by gsantner** [comment]
To be honest I would suggest to make it rather simple: always scan the filename at search too

**Comment by adelobosko** [comment]
Only i think need to rename the variable "count" in the commit [2cd2279](https://github.com/gsantner/markor/pull/1995/commits/2cd2279cb194e3b75c7b3f2d6f27e68848666e0a)

**Comment by harshad1** [comment]
@gsantner Making this so we don't need an option and always search filename too.

**Comment by gsantner** [comment]
Of course if you want to, just a suggestion. If it's preffered to be a choice you can keep it^^

**Comment by harshad1** [comment]
I actually can't think of any scenario where I want to search content but _not_ filename. I would rather make a simpler PR and add functionality later if I feel the need.

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java [hunk 3]
```diff
@@ -97,7 +97,9 @@ public static FileSearchEngine.QueueSearchFilesTask queueFileSearch(Activity act
     public static class QueueSearchFilesTask extends AsyncTask<Void, Integer, List<FitFile>> {
         private final SearchOptions _config;
         private final GsCallback.a1<List<FitFile>> _callback;
-        private final Pattern _regex;
+
+        // _matcher.reset() is _not_ thread safe. Will need alternate approach when we make search parallel
+        private final Matcher _matcher;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java:102]: Does adding a synchronized(object) around the two reset calls solve the problem?

**Review by gsantner — DISMISSED** [review]
Looks OK, do you still work on it or is it ready from your side?

About the matcher synchronized, it's just a comment and feel free to check it out, but it doesn't hinder merging.

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java (hunk 3 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java:111]: There is no issue right now as this code is not parallel. In the future I think searching needs to be made faster and I just left this as a note for myself or another dev in the future.

**Comment by harshad1** [comment]
I pushed up a change to also search folder names. Clicking on these opens the folder in the main activity.

@adelobosko I also pushed up some cleanups. Please take a look at the changes again. Any comments will be valuable.

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchResultSelectorDialog.java [hunk 4]
```diff
@@ -287,21 +295,4 @@ public void registerDataSetObserver(DataSetObserver dataSetObserver) {
         public void unregisterDataSetObserver(DataSetObserver dataSetObserver) {
         }
     }
-
-    public static class GroupItemsInfo {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchResultSelectorDialog.java:291]: This class is nearly identical to FitFile. Just using FitFile in both places

File: app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java [hunk 5]
```diff
@@ -61,8 +66,8 @@ public class GsFileUtils {
     public final static SimpleDateFormat DATEFORMAT_IMG = new SimpleDateFormat("yyyyMMdd-HHmmss", INITIAL_LOCALE); //20190511-230845
 
     // Used on methods like copyFile(src, dst)
-    private static final int BUFFER_SIZE = 4096;
-    private static final GsHashMap<String, String> MIME_TYPE_CACHE = new GsHashMap<>();
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java:65]: This is just a path -> mime cache. IMO many thousands of entries will not affect performance. I think we can just use a ConcurrentHashMap and we will gain some speed and safety

File: app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java [hunk 6]
```diff
@@ -420,7 +425,7 @@ public static String relativePath(File src, File dest) {
     }
 
     @SuppressWarnings("StatementWithEmptyBody")
-    private final static GsCallback.s1<File> gatherMimeType = file -> {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java:423]: Made this a regular function

File: app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchResultSelectorDialog.java [hunk 7]
```diff
@@ -217,11 +226,15 @@ public View getGroupView(final int groupPosition, final boolean isExpanded, fina
...
-            final int iconResId = groupInfo.isDirectory || groupInfo.children.isEmpty() ? 0 : isExpanded
-                    ? R.drawable.ic_baseline_keyboard_arrow_up_24
-                    : R.drawable.ic_baseline_keyboard_arrow_down_24;
+            final int iconResId;
+            if (groupInfo.isDirectory || groupInfo.children.isEmpty()) {
+                iconResId = 0;
+            } else if (isExpanded) {
+                iconResId = R.drawable.ic_baseline_keyboard_arrow_up_24;
+            } else {
+                iconResId = R.drawable.ic_baseline_keyboard_arrow_down_24;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchResultSelectorDialog.java:235]: Right now we don't specify an icon for file name results. Would it be better to use a folder or file icon instead?

I am adding a trailing slash to folders in order to indicate that they are folders

Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchResultSelectorDialog.java:235]: Usability improvements are sure not wrong. There is also the bug with the bottom snackbar not dismissing, or continue searching wrongly.

**Review by gsantner — APPROVED** [review]
Thanks, works good

👍🏻

**Pull request merged**